source/randLIB.c

```diff
@@ -14,125 +14,153 @@
  * limitations under the License.
  */
 #include <stdint.h>
-#include <stdlib.h>
-#include <stdbool.h>
 #include <limits.h>
 #include "randLIB.h"
 #include "platform/arm_hal_random.h"
 
-#if ((RAND_MAX+1) & RAND_MAX) != 0
-#error "RAND_MAX isn't 2^n-1 :("
-#endif
-
 /**
- * This library is made for getting random numbers for Timing needs in protocols.
+ * This library is made for getting random numbers for timing needs in
+ * protocols, plus to generate dynamic ports, random IDs etc.
  *
  * **not safe to use for security or cryptographic operations.**
  *
+ * Base implementation is a pseudo-RNG, but may also use a system RNG.
+ * Replay of sequence by reseeding is not possible.
+ *
+ * Base pseudo-RNG is the xoroshiro128+ generator by Marsaglia, Blackman and
+ * Vigna:
+ *
+ * http://xoroshiro.di.unimi.it/
+ *
+ * Certainly not the fastest for 32-bit or smaller platforms, but speed
+ * is not critical. None of the long operations in the core are actually hard,
+ * unlike the divisions and multiplies in the utility functions below, where we
+ * do try to keep the operations narrow.
  */
 
+/* On some platforms, read from a system RNG, rather than use our own */
+/* RANDLIB_PRNG disables this and forces use of the PRNG (useful for test only?) */
+#ifndef RANDLIB_PRNG
+#ifdef __linux
+#define RANDOM_DEVICE "/dev/urandom"
+#endif
+#endif // RANDLIB_PRNG
+
+/* RAM usage - 16 bytes of state (or a FILE * pointer and underlying FILE, which
+ * will include a buffer) */
+#ifdef RANDOM_DEVICE
+#include <stdio.h>
+static FILE *random_file;
+#else
+static uint64_t state[2];
+#endif
+
+#ifdef RANDLIB_PRNG
+void randLIB_reset(void)
+{
+    state[0] = 0;
+    state[1] = 0;
+}
+#endif
+
+#ifndef RANDOM_DEVICE
+static inline uint64_t rol(uint64_t n, int bits)
+{
+    return (n << bits) | (n >> (64 - bits));
+}
+#endif
 
-/**
-  * \brief Init seed for Pseudo Random.
-  *
-  * \return None
-  *
-  */
 void randLIB_seed_random(void)
 {
-    uint32_t rand_seed;
+#ifdef RANDOM_DEVICE
```

Comment:
Could do.

Current API, which I wanted to preserve was that entropy was "pulled" by calling arm_random_seed_get(), and this change modifies it so that repeated calls to randLIB_seed_random() would pull more out (rather than reset).

Was then considering maybe putting in occasional extra calls to arm_random_seed_get(), but was a bit nervous that some platforms might not like having arm_random_seed_get() called after boot.

So, yes, a seed_add(uint64_t) might be worthwhile. Won't do anything on RANDOM_DEVICE type things, of course.